src/api/objects/ObjectAPI.js

```diff
@@ -20,337 +20,387 @@
-     *           epoch, at which this domain object was last modified
-     * @property {module:openmct.ObjectAPI~Identifier[]} [composition] if
-     *           present, this will be used by the default composition provider
-     *           to load domain objects
-     * @typedef DomainObject
-     * @memberof module:openmct
-     */
-
-    function hasAlreadyBeenPersisted(domainObject) {
-        return domainObject.persisted !== undefined
-            && domainObject.persisted === domainObject.modified;
+            domainObject.persisted = persistedTime;
+            provider.create(domainObject).then((response) => {
+                this.mutate(domainObject, 'persisted', persistedTime);
+                savedResolve(response);
+            });
+        } else {
+            domainObject.persisted = persistedTime;
+            this.mutate(domainObject, 'persisted', persistedTime);
+            result = provider.update(domainObject);
+        }
+    }
+
+    return result;
+};
+
+/**
+ * Add a root-level object.
+ * @param {module:openmct.ObjectAPI~Identifier|function} an array of
+ *        identifiers for root level objects, or a function that returns a
+ *        promise for an identifier or an array of root level objects.
+ * @method addRoot
+ * @memberof module:openmct.ObjectAPI#
+ */
+ObjectAPI.prototype.addRoot = function (key) {
+    this.rootRegistry.addRoot(key);
+};
+
+/**
+ * Modify a domain object.
+ * @param {module:openmct.DomainObject} object the object to mutate
+ * @param {string} path the property to modify
+ * @param {*} value the new value for this property
+ * @method mutate
+ * @memberof module:openmct.ObjectAPI#
+ */
+ObjectAPI.prototype.mutate = function (domainObject, path, value) {
+    if (!this.supportsMutation(domainObject)) {
+        throw `Error: Attempted to mutate immutable object ${domainObject.name}`;
+    }
+
+    if (domainObject.isMutable) {
+        domainObject.$set(path, value);
+    } else {
+        //Creating a temporary mutable domain object allows other mutable instances of the
+        //object to be kept in sync.
+        let mutableDomainObject = this._toMutable(domainObject);
+
+        //Mutate original object
+        MutableDomainObject.mutateObject(domainObject, path, value);
+
+        //Mutate temptary mutable object, in the process informing any other mutable instances
+        mutableDomainObject.$set(path, value);
+
+        //Destroy temporary mutable object
+        this.destroyMutable(mutableDomainObject);
+    }
+};
+
+ObjectAPI.prototype._toMutable = function (object) {
+    if (object.isMutable) {
+        return object;
+    } else {
+        return MutableDomainObject.createMutable(object, this.eventEmitter);
     }
+};
+
+ObjectAPI.prototype.supportsMutation = function (object) {
+    // Checking for mutability is a bit broken right now. This is an 80% solution,
+    // but does not work in some cases.
+    const type = this.typeRegistry.get(object.type);
```

Comment:
For which cases does checking for mutability not work?

Reply:
eg. using folders to group things in a telemetry dictionary. The folder type is create-able (it appears in the create menu), but it shouldn't be mutable in the case of a telemetry dictionary. Our API doesn't allow us to distinguish between these two things right now, and we have been using "creatability" as a proxy for this for a long time. We need a better solution, but it's out of scope for this change. Maybe object providers should provide a hint about whether the objects they provide are mutable? 
eg.
```JavaScript
isMutable(objectIdentifier) {
    return false; // Or implement some more complicated logic here. Maybe the object provider returns a mix of mutable and non-mutable objects? I dunno.
}
```

At least if it's behind an API function for now it will make changing the implementation a little easier later on.

Reply:
Would this fix that: https://github.com/nasa/openmct/issues/3612 ? @akhenry

Reply:
@jvigliotta Yes, it will address the example case.